std/typecons.d

```diff
@@ -6896,3 +6896,222 @@ unittest
+    }
+
+    /++
+     +
+     + $(D HeadConst!T) implicitly converts to an rvalue of type $(D T) through
+     + $(D AliasThis).
+     +/
+    inout(T) HeadConst_get() inout
+    {
+        return HeadConst_value;
+    }
+
+    /// Ditto
+    alias HeadConst_get this;
+
+    auto ref T opUnary(string op)()
+        if (__traits(compiles, mixin(op ~ "T.init")))
+    {
+        return mixin(op ~ "this.HeadConst_value");
+    }
+}
+
+/// Ditto
+template HeadConst(T)
+    if (is(T == const) || is(T == immutable))
+{
+    alias HeadConst = T;
+}
+
+/// Ditto
+HeadConst!T headConst(T)(T t)
+{
+    return HeadConst!T(t);
+}
+
+/// $(D HeadConst) can be used to create class references which cannot be rebound:
+pure nothrow @safe unittest
+{
+    static class A
+    {
+        int i;
+
+        this(int i) pure nothrow @nogc @safe
+        {
+            this.i = i;
+        }
+    }
+
+    auto a = headConst(new A(42));
+    assert(a.i == 42);
+
+    // a = new C(24); // Reassignment is illegal,
+    a.i = 24; // But fields are still mutable.
+
+    assert(a.i == 24);
+}
+
+/// $(D HeadConst) can also be used to create read-only data fields without using transitive immutability:
+pure nothrow @safe unittest
+{
+    static class A
+    {
+        int i;
+
+        this(int i) pure nothrow @nogc @safe
+        {
+            this.i = i;
+        }
+    }
+
+    static class B
+    {
+        HeadConst!A a;
+
+        this(A a) pure nothrow @nogc @safe
+        {
+            this.a = a; // Construction, thus allowed.
+        }
+    }
+
+    auto b = new B(new A(42));
+    assert(b.a.i == 42);
+
+    // b.a = new A(24); // Reassignment is illegal,
+    b.a.i = 24; // but `a` is still mutable.
+
+    assert(b.a.i == 24);
+}
+
+pure nothrow @safe unittest
+{
+    static class A { int i; }
+    static assert(!is(HeadConst!A == A));
+    static assert(is(HeadConst!(const A) == const A));
+    static assert(is(HeadConst!(immutable A) == immutable A));
+
+    HeadConst!A a = new A;
+    static assert(!__traits(compiles, a = new A));
+
+    static void foo(ref A a) pure nothrow @safe @nogc {}
+    static assert(!__traits(compiles, foo(a)));
+
+    assert(a.i == 0);
+    a.i = 42;
+    assert(a.i == 42);
+
+    HeadConst!int i = 42;
+    static assert(!__traits(compiles, i = 24));
+    static assert(!__traits(compiles, --i));
+    static assert(!__traits(compiles, ++i));
+    assert(i == 42);
+    int iCopy = i;
+    assert(iCopy == 42);
+    iCopy = -i; // non-mutating unary operators must work
+    assert(iCopy == -42);
+
+    static struct S
+    {
+        int i;
+
+        pure nothrow @safe @nogc:
+        this(int i){}
+        this(string s){}
+        this(int i, string s, float f){ this.i = i; }
+    }
+
+    HeadConst!S sint = 42;
+    HeadConst!S sstr = "foo";
+    static assert(!__traits(compiles, sint = sstr));
+
+    auto sboth = HeadConst!S(42, "foo", 3.14);
+    assert(sboth.i == 42);
+
+    sboth.i = 24;
+    assert(sboth.i == 24);
```

Comment:
I don't think preventing assignment for structs when fields can still be assigned to makes much sense.